ddl/ddl_api.go

```diff
@@ -898,33 +888,102 @@ func (d *ddl) AddColumn(ctx context.Context, ti ast.Ident, spec *ast.AlterTableS
 		return ErrTooLongIdent.Gen("too long column %s", colName)
 	}
 
+	return nil
+}
+
+func (d *ddl) BuildColumnWithDefValue(ctx context.Context, t table.Table, column *ast.ColumnDef) (*table.Column, error) {
 	// Ingore table constraints now, maybe return error later.
 	// We use length(t.Cols()) as the default offset firstly, we will change the
 	// column's offset later.
-	col, _, err = buildColumnAndConstraint(ctx, len(t.Cols()), spec.NewColumn)
+	col, _, err := buildColumnAndConstraint(ctx, len(t.Cols()), column)
 	if err != nil {
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
 	col.OriginDefaultValue = col.DefaultValue
 	if col.OriginDefaultValue == nil && mysql.HasNotNullFlag(col.Flag) {
 		zeroVal := table.GetZeroValue(col.ToInfo())
 		col.OriginDefaultValue, err = zeroVal.ToString()
 		if err != nil {
-			return errors.Trace(err)
+			return nil, errors.Trace(err)
 		}
 	}
 
 	if col.OriginDefaultValue == expression.CurrentTimestamp &&
 		(col.Tp == mysql.TypeTimestamp || col.Tp == mysql.TypeDatetime) {
 		col.OriginDefaultValue = time.Now().Format(types.TimeFormat)
 	}
+	return col, nil
+}
+
+func (d *ddl) AppendColumns(ctx context.Context, ti ast.Ident, spec *ast.AlterTableSpec) error {
+	is := d.infoHandle.Get()
+	schema, ok := is.SchemaByName(ti.Schema)
+	if !ok {
+		return errors.Trace(infoschema.ErrDatabaseNotExists)
+	}
+	t, err := is.TableByName(ti.Schema, ti.Name)
+	if err != nil {
+		return errors.Trace(infoschema.ErrTableNotExists)
+	}
+
+	newColumns := []*table.Column{}
+	if err = checkDuplicateColumn(spec.AppendedColumns); err != nil {
+		return errors.Trace(err)
+	}
+	for _, column := range spec.AppendedColumns {
+		err = d.ValidColumnDef(column, t)
+		if err != nil {
+			return err
+		}
+
+		col, err := d.BuildColumnWithDefValue(ctx, t, column)
+		if err != nil {
+			return err
+		}
+		newColumns = append(newColumns, col)
+	}
+
+	job := &model.Job{
+		SchemaID:   schema.ID,
+		TableID:    t.Meta().ID,
+		Type:       model.ActionAddColumns,
+		BinlogInfo: &model.HistoryInfo{},
+		Args:       []interface{}{newColumns, []*ast.ColumnPosition{}, 0},
```

Comment:
Why not pass column positions ? The multi-add-columns can also specify positions.